src/server/async/socket_server.cc

```diff
@@ -514,6 +521,42 @@ bool SocketConnection::doCreateBuffer(const json& root) {
   return false;
 }
 
+bool SocketConnection::doCreateGPUBuffer(json const& root) {
+  auto self(shared_from_this());
+
+  size_t size;
+  std::shared_ptr<Payload> object;
+  std::string message_out;
+
+  TRY_READ_REQUEST(ReadCreateGPUBufferRequest, root, size);
+  ObjectID object_id;
+  RESPONSE_ON_ERROR(gpu_bulk_store_->Create(size, object_id, object));
+  if(!object->IsGPU() || object->pointer == nullptr) {
+    LOG(INFO) << "invalid GPU memory pointer"<< std::endl;
+    return false;
+  }
+  cudaIpcMemHandle_t cuda_handle;
+  if(object->data_size > 0 && 
+     self->used_cuda_ipc_mem_handle_.find(object->pointer) == 
+     self->used_cuda_ipc_mem_handle_.find(object->pointer)) {
+    cudaError_t result =cudaIpcGetMemHandle((cudaIpcMemHandle_t *)&cuda_handle, 
+                                            object->pointer);
+    if(result != cudaSuccess){
+      LOG(ERROR) << "cudaIpcGetMemHandle error: " << cudaGetErrorString(result) << std::endl;
+      return false;
+    }
+    self->used_cuda_ipc_mem_handle_[object->pointer] = cuda_handle;
```

Comment:
why do we need to maintain the `used_cuda_ipc_mem_handle`?

Reply:
In case that different clients request the same memory on GPU, then cudaIpcGetMemHandle can be avoided